src/custom/constants/networks.ts

```diff
@@ -28,3 +38,142 @@ export const RPC_URLS: { [key in SupportedChainId]: string } = {
+  [SupportedChainId.MAINNET]: {
+    chainId: SupportedChainId.MAINNET,
+    name: 'homestead',
+  },
+  [SupportedChainId.GOERLI]: {
+    chainId: SupportedChainId.GOERLI,
+    name: 'goerli',
+  },
+  [SupportedChainId.GNOSIS_CHAIN]: {
+    chainId: SupportedChainId.GNOSIS_CHAIN,
+    name: 'gnosis',
+  },
+}
+
+/**
+ * We use this class for type inference.
+ *
+ * UrlJsonRpcProvider is an abstract class, therefore we have trouble inferring the type of its constructor.
+ * However, as all of the RPC providers we use extend UrlJsonRpcProvider, we can use this class to infer the type of the constructor, by turning it into a concrete one.
+ */
+class BaseUrlJsonRpcProvider extends UrlJsonRpcProvider {}
+type ProviderClass = Newable<typeof BaseUrlJsonRpcProvider>
+
+/**
+ * UrlJsonRpcProvider instances do not have a getSigner implementation, however we need it.
+ *
+ * As they are all based on JsonRpcProvider, we can use JsonRpcProvider's getSigner implementation.
+ *
+ * Why not use JsonRpcProvider directly? Because it does not have specific API provider optimisations,
+ * and it does not automatically support community resources.
+ *
+ * Why not instantiate and return a JsonRpcSigner? Because it has a constructor guard that prevents it from being instantiated directly.
+ * Intention with this guard is to prevent users from accidentally using a JsonRpcSigner with another provider. Which is not possible here.
+ */
+class InfuraProvider extends EthersInfuraProvider {}
+InfuraProvider.prototype.getSigner = JsonRpcProvider.prototype.getSigner
+class CloudflareProvider extends EthersCloudflareProvider {}
+CloudflareProvider.prototype.getSigner = JsonRpcProvider.prototype.getSigner
+class AlchemyProvider extends EthersAlchemyProvider {}
+AlchemyProvider.prototype.getSigner = JsonRpcProvider.prototype.getSigner
+class PocketProvider extends EthersPocketProvider {}
+PocketProvider.prototype.getSigner = JsonRpcProvider.prototype.getSigner
+class AnkrProvider extends EthersAnkrProvider {}
+AnkrProvider.prototype.getSigner = JsonRpcProvider.prototype.getSigner
+
+// List of public providers that don't require an API key, per chain support.
+const PUBLIC_PROVIDER_LIST: Record<SupportedChainId, ProviderClass[]> = {
+  [SupportedChainId.MAINNET]: [InfuraProvider, CloudflareProvider, AlchemyProvider, PocketProvider, AnkrProvider],
+  [SupportedChainId.GOERLI]: [InfuraProvider, AlchemyProvider, PocketProvider, AnkrProvider],
+  [SupportedChainId.GNOSIS_CHAIN]: [],
+}
+
+const RPC_TYPE: { [key in SupportedChainId]: 'infura' | 'custom' } = {
+  [SupportedChainId.MAINNET]: 'infura',
+  [SupportedChainId.GOERLI]: 'infura',
+  [SupportedChainId.GNOSIS_CHAIN]: 'custom',
+}
+
+function getProvider(chainId: SupportedChainId): JsonRpcProvider[] {
+  const result: JsonRpcProvider[] = []
+
+  if (window.ethereum) {
+    result.push(new Web3Provider(window.ethereum, RPC_CONFIG[chainId]))
+  }
+
+  result.pop()
```

Comment:
Why do we do `pop`? Looks like it just cancel 105 line

Reply:
Oh no! My fault - it was a bit too late I guess. It was a debug line to see everything else working as expected. Removed it.